### PR TITLE
[AIRFLOW-5631] Change way of running GCP system tests

### PR DESCRIPTION
**Pull request by turbaszek**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5631

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
This PR proposes a new way of running GCP related system tests.
It uses SystemTests base class and authentication is provided by a
context manager thus it's easier to understand what's going on.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: airflow/gcp/hooks/base.py [hunk 1]
```diff
@@ -323,8 +341,63 @@ def provide_gcp_credential_file_as_context(self):
...
+        scopes: Optional[Sequence[str]] = None,
+        project_id: Optional[str] = None,
+        gcp_conn_id: str = 'google_cloud_default'
+    ) -> str:
+        """
+        Builds a variable that can be used as `AIRFLOW_CONN_GOOGLE_CLOUD_DEFAULT` with provided service key,
+        scopes and project id.
+        """
+        conn = "google-cloud-platform://?"
+        extras = "extra__{}".format(gcp_conn_id)
```
Comment (mik-laj) [inline airflow/gcp/hooks/base.py]: It's will probable doesn't work. This text is related to UI.

File: airflow/gcp/hooks/base.py [hunk 2]
```diff
@@ -323,8 +341,63 @@ def provide_gcp_credential_file_as_context(self):
...
+        scopes = scopes or _DEFAULT_SCOPES
+        project_id = project_id or os.environ.get(PROJECT)
+
+        if key_file_path:
+            conn += "{}__key_path={}&".format(extras, key_file_path)
```
Comment (mik-laj) [inline airflow/gcp/hooks/base.py]: I don't like creating a URL using string concatenation. In my opinion, https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode is a better solution.

File: airflow/gcp/hooks/base.py [hunk 3]
```diff
@@ -294,6 +295,26 @@ def wrapper(self: GoogleCloudBaseHook, *args, **kwargs) -> RT:
                 return func(self, *args, **kwargs)
         return wrapper
 
+    @staticmethod
+    def validate_key(key_path: str) -> None:
```
Comment (mik-laj) [inline airflow/gcp/hooks/base.py]: I'm afraid this should no longer be part of the public part of hook interface.

File: airflow/gcp/hooks/base.py [hunk 4]
```diff
@@ -323,8 +341,63 @@ def provide_gcp_credential_file_as_context(self):
...
-                        del os.environ[CREDENTIALS]
-                else:
-                    os.environ[CREDENTIALS] = current_env_state
+                self.restore_env_variable(current_env_state)
+
+    @staticmethod
+    def build_gcp_conn(
+        key_file_path: Optional[str] = None,
+        scopes: Optional[Sequence[str]] = None,
+        project_id: Optional[str] = None,
```
Comment (mik-laj) [inline airflow/gcp/hooks/base.py]: It's a hook method, so it should read a project_id from hook also.

File: airflow/gcp/hooks/base.py [hunk 5]
```diff
@@ -323,8 +341,63 @@ def provide_gcp_credential_file_as_context(self):
...
+        if scopes:
+            conn += "{}__scope={}&".format(extras, ",".join(scopes))
+        if project_id:
+            conn += "{}__project={}".format(extras, project_id)
+
+        return conn
+
+    @staticmethod
+    def _resolve_gcp_keys_path() -> str:
+        path = os.environ.get('GCP_CONFIG_DIR', "/config")
```
Comment (mik-laj) [inline airflow/gcp/hooks/base.py]: Where is it used in Airflow's core?

File: airflow/gcp/hooks/base.py (hunk 4 above)
Comment (turbaszek) [reply airflow/gcp/hooks/base.py]: My idea was to keep this method stateless. Part of the reason was the fact it's used in `provide_gcp_credentials_from_service_key` which is also a static method.

Comment (mik-laj) [reply airflow/gcp/hooks/base.py]: We discussed it offline

File: airflow/gcp/hooks/base.py (hunk 5 above)
Comment (turbaszek) [reply airflow/gcp/hooks/base.py]: What do you mean?

File: airflow/gcp/hooks/base.py (hunk 1 above)
Comment (turbaszek) [reply airflow/gcp/hooks/base.py]: What do you mean by not working?

File: airflow/gcp/hooks/base.py (hunk 5 above)
Comment (mik-laj) [reply airflow/gcp/hooks/base.py]: This is only necessary in the test package, so it should be in the test package.

File: airflow/gcp/hooks/base.py (hunk 1 above)
Comment (mik-laj) [reply airflow/gcp/hooks/base.py]: The program will not set the appropriate value. Look at the documentation.
https://airflow.apache.org/howto/connection/gcp.html
and in source code:
https://github.com/apache/airflow/blob/master/airflow/www/static/js/connection_form.js#L29

Comment (turbaszek) [reply airflow/gcp/hooks/base.py]: I think I still miss something. This works at least when I run system tests.

Comment (mik-laj) [reply airflow/gcp/hooks/base.py]: This only works with the default connection ID. https://github.com/apache/airflow/blob/master/airflow/gcp/hooks/base.py#L163-L174

File: airflow/gcp/hooks/base.py (hunk 2 above)
Comment (turbaszek) [reply airflow/gcp/hooks/base.py]: Good point!

**Comment by mik-laj** [comment]
Travis is sad. Can you fix it?

File: airflow/utils/log/colored_log.py [hunk 6]
```diff
@@ -30,7 +30,7 @@
 ARGS = {"attrs": ["bold"]}
 
 DEFAULT_COLORS = {
-    "DEBUG": "red",
+    "DEBUG": "green",
```
Comment (mik-laj) [inline airflow/utils/log/colored_log.py]: Is it unrelated changes?

File: tests/gcp/operators/test_cloud_sql_system.py [hunk 7]
```diff
@@ -17,42 +17,34 @@
...
-        self.gcp_authenticator.gcp_authenticate()
-        try:
-            SQL_QUERY_TEST_HELPER.delete_instances(instance_suffix="-failover-replica")
-            SQL_QUERY_TEST_HELPER.delete_instances(instance_suffix="-read-replica")
-            SQL_QUERY_TEST_HELPER.delete_instances()
-            SQL_QUERY_TEST_HELPER.delete_instances(instance_suffix="2")
-            SQL_QUERY_TEST_HELPER.delete_service_account_acls()
-        finally:
-            self.gcp_authenticator.gcp_revoke_authentication()
+        SQL_QUERY_TEST_HELPER.delete_instances(instance_suffix="-failover-replica")
```
Comment (mik-laj) [inline tests/gcp/operators/test_cloud_sql_system.py:37]: In other places, the helper is set as a class property. Do you have recommendations about it?

Comment (turbaszek) [reply tests/gcp/operators/test_cloud_sql_system.py:37]: I think the helpers need some more discussion. Many of them performs same thing like create a bucket or delete a bucket, so maybe it's worth to rethink our overall approach. If we are going to migrate to pytest then helpers should be replaced by fixtures.

File: airflow/gcp/utils/credentials_provider.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+AIRFLOW_CONN_GOOGLE_CLOUD_DEFAULT = "AIRFLOW_CONN_GOOGLE_CLOUD_DEFAULT"
+
+
+def build_gcp_conn(
+    key_file_path: Optional[str] = None,
+    scopes: Optional[Sequence[str]] = None,
+    project_id: Optional[str] = None,
+) -> str:
+    """
+    Builds a variable that can be used as `AIRFLOW_CONN_GOOGLE_CLOUD_DEFAULT` with provided service key,
```
Comment (mik-laj) [inline airflow/gcp/utils/credentials_provider.py]:
```suggestion
    Builds a variable that can be used as ``AIRFLOW_CONN_GOOGLE_CLOUD_DEFAULT`` with provided service key,
```

File: airflow/gcp/utils/credentials_provider.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+
+@contextmanager
+def provide_gcp_connection(
+    key_file_path: Optional[str] = None,
+    scopes: Optional[Sequence] = None,
+    project_id: Optional[str] = None,
+):
+    """
+    Context manager that provides a temporary value of AIRFLOW_CONN_GOOGLE_CLOUD_DEFAULT
```
Comment (mik-laj) [inline airflow/gcp/utils/credentials_provider.py:139]:
```suggestion
    Context manager that provides a temporary value of ``AIRFLOW_CONN_GOOGLE_CLOUD_DEFAULT``
```

File: airflow/gcp/utils/credentials_provider.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+
+
+@contextmanager
+def provide_gcp_conn_and_credentials(
+    key_file_path: Optional[str] = None,
+    scopes: Optional[Sequence] = None,
+    project_id: Optional[str] = None,
+):
+    """
+    Context manager that provides both:
```
Comment (mik-laj) [inline airflow/gcp/utils/credentials_provider.py:170]:
```suggestion
    Context manager that provides both:
    
```

File: tests/gcp/operators/test_cloud_sql_system.py (hunk 7 above)
Comment (mik-laj) [reply tests/gcp/operators/test_cloud_sql_system.py:37]: Ok. We can do it in separate PR.

File: tests/test_utils/gcp_system_helpers.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+
+
+def provide_gcp_context(
+    key_file_path: Optional[str] = None,
+    scopes: Optional[Sequence] = None,
+    project_id: Optional[str] = None,
+):
+    """
+    Context manager that provides both:
+    - GCP credentials for application supporting `Application Default Credentials (ADC)
```
Comment (mik-laj) [inline tests/test_utils/gcp_system_helpers.py:168]:
```suggestion

    - GCP credentials for application supporting `Application Default Credentials (ADC)
```

File: tests/test_utils/gcp_system_helpers.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        return unittest.skip(LOCAL_EXECUTOR_WARNING.format(POSTGRES_LOCAL_EXECUTOR))
+
+    return lambda cls: cls
+
+
+def resolve_full_gcp_key_path(key: str) -> str:
+    """
+    Returns path full path to provided GCP key.
+
+    :param key: Name of the GCP key, for example "my_service.json"
```
Comment (mik-laj) [inline tests/test_utils/gcp_system_helpers.py]:
```suggestion
    :param key: Name of the GCP key, for example: ``my_service.json``
```

File: tests/test_utils/gcp_system_helpers.py [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+def provide_gcp_context(
+    key_file_path: Optional[str] = None,
+    scopes: Optional[Sequence] = None,
+    project_id: Optional[str] = None,
+):
+    """
+    Context manager that provides both:
+    - GCP credentials for application supporting `Application Default Credentials (ADC)
+    strategy <https://cloud.google.com/docs/authentication/production>`__.
+    - temporary value of AIRFLOW_CONN_GOOGLE_CLOUD_DEFAULT connection
```
Comment (mik-laj) [inline tests/test_utils/gcp_system_helpers.py]:
```suggestion
    - temporary value of ``AIRFLOW_CONN_GOOGLE_CLOUD_DEFAULT`` connection
```

File: tests/test_utils/gcp_system_helpers.py [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+
+    Moreover it resolves full path to service keys so user can pass 'myservice.json`
+    as `key_file_path`.
```
Comment (mik-laj) [inline tests/test_utils/gcp_system_helpers.py]:
```suggestion
    as ``key_file_path``.
```

File: tests/gcp/utils/base_gcp_system_test_case.py [hunk 15]
```diff
@@ -241,11 +238,9 @@ def _check_local_executor_setup():
 Please set AIRFLOW_CONFIG variable to '{}'
 and make sure you have a Postgres server running locally and
 airflow/airflow.db database created.
```
Comment (mik-laj) [inline tests/gcp/utils/base_gcp_system_test_case.py]: Is this class still used?

**Review by mik-laj — APPROVED** [review]
Small nits.

File: tests/gcp/utils/base_gcp_system_test_case.py (hunk 15 above)
Comment (turbaszek) [reply tests/gcp/utils/base_gcp_system_test_case.py]: No, I should remove it ✅

**Review by potiuk — APPROVED** [review]

File: tests/gcp/operators/test_bigquery_system.py [hunk 16]
```diff
@@ -17,40 +17,32 @@
...
 from tests.gcp.operators.test_bigquery_system_helper import GCPBigQueryTestHelper
-from tests.gcp.utils.base_gcp_system_test_case import SKIP_TEST_WARNING, TestDagGcpSystem
 from tests.gcp.utils.gcp_authenticator import GCP_BIGQUERY_KEY
+from tests.test_utils.gcp_system_helpers import GCP_DAG_FOLDER, provide_gcp_context, skip_gcp_system
+from tests.test_utils.system_tests_class import SystemTest
 
 
-@unittest.skipIf(TestDagGcpSystem.skip_check(GCP_BIGQUERY_KEY), SKIP_TEST_WARNING)
-class BigQueryExampleDagsSystemTest(TestDagGcpSystem):
+@skip_gcp_system(GCP_BIGQUERY_KEY, require_local_executor=True)
```
Comment (TobKed) [inline tests/gcp/operators/test_bigquery_system.py:27]: This parameters should not have `require_local_executor` set to `True` I think. I have problem with running this test.

Comment (turbaszek) [reply tests/gcp/operators/test_bigquery_system.py:27]: Yes, I agree. Can you make a PR with fix because I am on holidays?

Comment (TobKed) [reply tests/gcp/operators/test_bigquery_system.py:27]: Sorry for late response. I will try to fix it but I have limited time due to my holidays :)
